File: apps/qa-playground-flutter/test/components/checkbox/checkbox_a11y_test.dart
/// Checkbox accessibility QA tests — RN `CheckboxA11y.test.ts` + widget semantics.
library;

import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:ui_flutter/widgets/one_ui_checkbox.dart';
import 'package:ui_flutter/widgets/one_ui_checkbox_a11y.dart';
import 'package:ui_flutter/widgets/one_ui_checkbox_types.dart';

import '../../support/components/checkbox_harness.dart';

void main() {
  group('[a11y] resolveCheckboxAccessibilityLabel — RN label fallback chain', () {
    test('[a11y] prefers semanticsLabel over ariaLabel and visible label', () {
      expect(
        resolveCheckboxAccessibilityLabel(
          semanticsLabel: 'Screen reader',
          ariaLabel: 'Aria',
          label: 'Visible',
        ),
        'Screen reader',
      );
    });

    test('[a11y] prefers ariaLabel over visible label', () {
      expect(
        resolveCheckboxAccessibilityLabel(
          ariaLabel: 'Aria only',
          label: 'Visible',
        ),
        'Aria only',
      );
    });

    test('[a11y] falls back to description when label omitted', () {
      expect(
        resolveCheckboxAccessibilityLabel(
          description: 'Weekly digest only.',
        ),
        'Weekly digest only.',
      );
    });
  });

  group('[a11y] getCheckboxAccessibilityProps — RN CheckboxA11y.test.ts', () {
    test('[a11y] exposes checked=true when selected', () {
      final a11y = getCheckboxAccessibilityProps(
        ariaLabel: 'Accept terms',
        isDisabled: false,
        isReadOnly: false,
        isChecked: true,
        isIndeterminate: false,
      );
      expect(a11y.exposeControl, isTrue);
      expect(a11y.label, 'Accept terms');
      expect(a11y.checked, isTrue);
      expect(a11y.canTap, isTrue);
    });

    test('[a11y] reports mixed state for indeterminate', () {
      final a11y = getCheckboxAccessibilityProps(
        ariaLabel: 'Mixed',
        isDisabled: false,
        isReadOnly: false,
        isChecked: false,
        isIndeterminate: true,
      );
      expect(a11y.checked, 'mixed');
    });

    test('[a11y] forwards aria-describedby as describedBy', () {
      final a11y = getCheckboxAccessibilityProps(
        ariaLabel: 'Box',
        ariaDescribedby: 'help-1',
        isDisabled: false,
        isReadOnly: false,
        isChecked: false,
        isIndeterminate: false,
      );
      expect(a11y.describedBy, 'help-1');
    });

    test('[a11y] marks disabled state and blocks tap', () {
      final a11y = getCheckboxAccessibilityProps(
        ariaLabel: 'Box',
        isDisabled: true,
        isReadOnly: false,
        isChecked: false,
        isIndeterminate: false,
      );
      expect(a11y.canTap, isFalse);
    });

    test('[a11y] readOnly allows exposure but blocks tap', () {
      final a11y = getCheckboxAccessibilityProps(
        label: 'Locked',
        isDisabled: false,
        isReadOnly: true,
        isChecked: true,
        isIndeterminate: false,
      );
      expect(a11y.exposeControl, isTrue);
      expect(a11y.canTap, isFalse);
    });

    test('[a11y] aria-hidden collapses exposure', () {
      final a11y = getCheckboxAccessibilityProps(
        ariaLabel: 'Hidden',
        ariaHidden: true,
        isDisabled: false,
        isReadOnly: false,
        isChecked: false,
        isIndeterminate: false,
      );
      expect(a11y.hidden, isTrue);
      expect(a11y.exposeControl, isFalse);
    });
  });

  group('[a11y] resolveOneUiCheckboxState — RN useCheckboxState', () {
    test('[a11y] defaults size=m, appearance=secondary, enabled', () {
      final state = resolveOneUiCheckboxState(isChecked: false);
      expect(state.resolvedSize, 'm');
      expect(state.resolvedAppearance, 'secondary');
      expect(state.isDisabled, isFalse);
      expect(state.isReadOnly, isFalse);
    });

    test('[a11y] canonicalises legacy size aliases', () {
      expect(resolveOneUiCheckboxState(size: 'small', isChecked: false).resolvedSize, 's');
      expect(resolveOneUiCheckboxState(size: 'medium', isChecked: false).resolvedSize, 'm');
      expect(resolveOneUiCheckboxState(size: 'large', isChecked: false).resolvedSize, 'l');
    });
  });

  group('[a11y] Checkbox — role and name (RN a11y parity)', () {
    testWidgetsAllPlatforms('[a11y] checkbox semantics expose checked state when unselected', (tester) async {
      await pumpCheckboxQaHarness(tester, OneUiCheckbox(label: 'Option', checked: false));
      withSemanticsHandle(tester, () {
        final d = checkboxSemanticsData(tester, 'Option', checked: false);
        expect(d.hasFlag(SemanticsFlag.hasCheckedState), isTrue);
        expect(d.hasFlag(SemanticsFlag.isChecked), isFalse);
      });
    });

    testWidgetsAllPlatforms('[a11y] label prop becomes accessible name', (tester) async {
      await pumpCheckboxQaHarness(tester, OneUiCheckbox(label: 'Accept terms'));
      withSemanticsHandle(tester, () {
        expect(checkboxSemanticsLabel('Accept terms'), findsOneWidget);
      });
    });

    testWidgetsAllPlatforms('[a11y] readOnly stays enabled in semantics (RN parity)', (tester) async {
      await pumpCheckboxQaHarness(tester, OneUiCheckbox(readOnly: true, label: 'Locked'));
      expectCheckboxReadOnlyEnabled(tester, 'Locked');
    });

    testWidgetsAllPlatforms('[a11y] disabled marks not enabled', (tester) async {
      await pumpCheckboxQaHarness(tester, OneUiCheckbox(disabled: true, label: 'Off'));
      expectCheckboxDisabled(tester, 'Off');
    });
  });

  group('[a11y] OneUiCheckbox — widget semantics', () {
    testWidgetsAllPlatforms('[a11y] description-only uses description as accessible name', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(description: 'Weekly digest only.'),
      );
      expect(find.text('Weekly digest only.'), findsOneWidget);
      withSemanticsHandle(tester, () {
        expectCheckboxChecked(tester, 'Weekly digest only.', checked: false);
      });
    });

    testWidgetsAllPlatforms('[a11y] ariaLabel overrides visible label in semantics', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Visible',
          ariaLabel: 'From aria',
        ),
      );
      withSemanticsHandle(tester, () {
        expect(checkboxSemanticsLabel('From aria'), findsOneWidget);
        expect(checkboxSemanticsLabel('Visible'), findsNothing);
      });
    });

    testWidgetsAllPlatforms('[a11y] renders without visible label', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(ariaLabel: 'Standalone'),
      );
      withSemanticsHandle(tester, () {
        expect(checkboxSemanticsLabel('Standalone'), findsOneWidget);
      });
    });

    testWidgetsAllPlatforms('[a11y] checked semantics flip on tap', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Toggle',
          defaultChecked: false,
        ),
      );

      expectCheckboxChecked(tester, 'Toggle', checked: false);
      await tester.tap(find.text('Toggle'));
      await tester.pumpAndSettle();
      expectCheckboxChecked(tester, 'Toggle', checked: true);
    });

    testWidgetsAllPlatforms('[a11y] indeterminate exposes mixed semantics', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Select all',
          indeterminate: true,
        ),
      );
      expectCheckboxMixed(tester, 'Select all');
    });

    testWidgetsAllPlatforms('[a11y] disabled blocks toggle and semantics', (tester) async {
      var changed = false;
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Disabled',
          disabled: true,
          onCheckedChange: (_) => changed = true,
        ),
      );

      expectCheckboxDisabled(tester, 'Disabled');
      await tester.tap(find.text('Disabled'));
      await tester.pumpAndSettle();
      expect(changed, isFalse);
    });

    testWidgetsAllPlatforms('[a11y] readOnly stays enabled but cannot toggle', (tester) async {
      var changed = false;
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Locked',
          readOnly: true,
          defaultChecked: false,
          onCheckedChange: (_) => changed = true,
        ),
      );

      expectCheckboxReadOnlyEnabled(tester, 'Locked');
      await tester.tap(find.text('Locked'));
      await tester.pumpAndSettle();
      expect(changed, isFalse);
      expectCheckboxChecked(tester, 'Locked', checked: false);
    });

    testWidgetsAllPlatforms('[a11y] aria-hidden excludes control from semantics tree', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Hidden',
          ariaHidden: true,
        ),
      );
      withSemanticsHandle(tester, () {
        expect(checkboxSemanticsLabel('Hidden'), findsNothing);
      });
    });

    testWidgetsAllPlatforms('[a11y] accessibilityHint surfaces in semantics', (tester) async {
      await pumpCheckboxQaHarness(
        tester,
        OneUiCheckbox(
          label: 'Hinted',
          accessibilityHint: 'Toggles the option',
        ),
      );
      withSemanticsHandle(tester, () {
        final data = checkboxSemanticsData(tester, 'Hinted');
        expect(data.hint, 'Toggles the option');
      });
    });
  });
}

File: packages/ui-native/src/components/Checkbox/CheckboxA11y.test.ts
import { describe, expect, it } from 'vitest';
import {
  getCheckboxAccessibilityProps,
  resolveSize,
  useCheckboxState,
} from './interface';

describe('checkboxA11y', () => {
  it('exposes role=checkbox with state.checked=true when selected', () => {
    const a11y = getCheckboxAccessibilityProps(
      { 'aria-label': 'Accept terms' },
      { isDisabled: false, isReadOnly: false, isSelected: true, isIndeterminate: false },
    );
    expect(a11y.accessible).toBe(true);
    expect(a11y.accessibilityRole).toBe('checkbox');
    expect(a11y.accessibilityLabel).toBe('Accept terms');
    expect(a11y.accessibilityState.checked).toBe(true);
    expect(a11y.accessibilityState.disabled).toBe(false);
  });

  it('reports mixed state for indeterminate', () => {
    const a11y = getCheckboxAccessibilityProps(
      { 'aria-label': 'Mixed' },
      { isDisabled: false, isReadOnly: false, isSelected: false, isIndeterminate: true },
    );
    expect(a11y.accessibilityState.checked).toBe('mixed');
  });

  it('falls back to the visible label when aria-label is omitted', () => {
    const a11y = getCheckboxAccessibilityProps(
      { label: 'Subscribe to updates' },
      { isDisabled: false, isReadOnly: false, isSelected: false, isIndeterminate: false },
    );
    expect(a11y.accessibilityLabel).toBe('Subscribe to updates');
  });

  it('forwards aria-describedby as accessibilityLabelledBy', () => {
    const a11y = getCheckboxAccessibilityProps(
      { 'aria-label': 'Box', 'aria-describedby': 'help-1' },
      { isDisabled: false, isReadOnly: false, isSelected: false, isIndeterminate: false },
    );
    expect(a11y.accessibilityLabelledBy).toBe('help-1');
  });

  it('hides the subtree when aria-hidden=true', () => {
    const a11y = getCheckboxAccessibilityProps(
      { 'aria-label': 'Box', 'aria-hidden': true },
      { isDisabled: false, isReadOnly: false, isSelected: false, isIndeterminate: false },
    );
    expect(a11y.accessible).toBe(false);
    expect(a11y.accessibilityElementsHidden).toBe(true);
    expect(a11y.importantForAccessibility).toBe('no-hide-descendants');
  });

  it('passes accessibilityHint through unchanged', () => {
    const a11y = getCheckboxAccessibilityProps(
      { 'aria-label': 'Box', accessibilityHint: 'Toggles the option' },
      { isDisabled: false, isReadOnly: false, isSelected: false, isIndeterminate: false },
    );
    expect(a11y.accessibilityHint).toBe('Toggles the option');
  });

  it('marks disabled state', () => {
    const a11y = getCheckboxAccessibilityProps(
      { 'aria-label': 'Box' },
      { isDisabled: true, isReadOnly: false, isSelected: false, isIndeterminate: false },
    );
    expect(a11y.accessibilityState.disabled).toBe(true);
  });
});

describe('useCheckboxState', () => {
  it('defaults size=m, appearance=secondary (auto), disabled=false', () => {
    const state = useCheckboxState({});
    expect(state.resolvedSize).toBe('m');
    expect(state.resolvedAppearance).toBe('secondary');
    expect(state.isDisabled).toBe(false);
    expect(state.isReadOnly).toBe(false);
  });

  it('resolves auto appearance to secondary (matches web)', () => {
    const state = useCheckboxState({ appearance: 'auto' });
    expect(state.resolvedAppearance).toBe('secondary');
  });

  it('keeps explicit appearance', () => {
    const state = useCheckboxState({ appearance: 'positive' });
    expect(state.resolvedAppearance).toBe('positive');
  });

  it('canonicalises legacy size aliases', () => {
    expect(useCheckboxState({ size: 'small' }).resolvedSize).toBe('s');
    expect(useCheckboxState({ size: 'medium' }).resolvedSize).toBe('m');
    expect(useCheckboxState({ size: 'large' }).resolvedSize).toBe('l');
  });

  it('emits data attributes mirroring web (.module.css selectors)', () => {
    const unchecked = useCheckboxState({ size: 'l', appearance: 'negative' });
    expect(unchecked.dataAttrs).toEqual({
      'data-size': 'l',
      'data-appearance': 'negative',
      'data-unselected': '',
    });

    const sel = useCheckboxState({ size: 'm', selected: true });
    expect(sel.dataAttrs).toMatchObject({ 'data-selected': '' });

    const indeterminate = useCheckboxState({ size: 's', indeterminate: true });
    expect(indeterminate.dataAttrs).toMatchObject({ 'data-indeterminate': '' });

    const ro = useCheckboxState({ readOnly: true, selected: true });
    expect(ro.dataAttrs).toMatchObject({ 'data-readonly': '', 'data-selected': '' });
  });

  it('forwards aria props on disabled / read-only', () => {
    expect(getCheckboxAccessibilityProps({}, { isDisabled: true, isReadOnly: false, isSelected: false, isIndeterminate: false })['aria-disabled']).toBe(true);
    expect(getCheckboxAccessibilityProps({}, { isDisabled: false, isReadOnly: true, isSelected: false, isIndeterminate: false })['aria-readonly']).toBe(true);
  });
});

describe('resolveSize', () => {
  it.each([
    ['s', 's'],
    ['m', 'm'],
    ['l', 'l'],
    ['small', 's'],
    ['medium', 'm'],
    ['large', 'l'],
  ] as const)('canonicalises %s → %s', (input, expected) => {
    expect(resolveSize(input)).toBe(expected);
  });
});
